test-data/unit/check-type-aliases.test

```diff
@@ -1065,3 +1065,99 @@ def eval(e: Expr) -> int:
+
+T1 = T2 = TypeAliasType("T", int)
+t1: T1  # E: Variable "__main__.T1" is not valid as a type \
+        # N: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
+
+T3 = TypeAliasType("T3", -1)
+t3: T3  # E: Variable "__main__.T3" is not valid as a type \
+        # N: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
+[builtins fixtures/tuple.pyi]
+
+[case testTypeAliasTypeGeneric]
+from typing_extensions import TypeAliasType, TypeVarTuple, ParamSpec
+from typing import Callable, Dict, TypeVar, Tuple, Unpack
+
+K = TypeVar('K')
+V = TypeVar('V')
+Ts = TypeVarTuple("Ts")
+Ts1 = TypeVarTuple("Ts1")
+P = ParamSpec("P")
+
+TestType = TypeAliasType("TestType", Dict[K, V], type_params=(K, V))
+x: TestType[int, str] = {1: 'a'}
+y: TestType[str, int] = {'a': 1}
+z: TestType[str, int] = {1: 'a'}  # E: Dict entry 0 has incompatible type "int": "str"; expected "str": "int"
+
+InvertedDict = TypeAliasType("InvertedDict", Dict[K, V], type_params=(V, K))
+xi: InvertedDict[str, int] = {1: 'a'}
+yi: InvertedDict[str, int] = {'a': 1}  # E: Dict entry 0 has incompatible type "str": "int"; expected "int": "str"
+zi: InvertedDict[int, str] = {1: 'a'}  # E: Dict entry 0 has incompatible type "int": "str"; expected "str": "int"
+reveal_type(xi)  # N: Revealed type is "builtins.dict[builtins.int, builtins.str]"
+
+VariadicAlias1 = TypeAliasType("VariadicAlias1", Tuple[Unpack[Ts]], type_params=(Ts,))
+VariadicAlias2 = TypeAliasType("VariadicAlias2", Tuple[Unpack[Ts], K], type_params=(Ts, K))
+xv: VariadicAlias1[int, str] = (1, 'a')
+yv: VariadicAlias1[str, int] = (1, 'a')  # E: Incompatible types in assignment (expression has type "Tuple[int, str]", variable has type "Tuple[str, int]")
+zv: VariadicAlias2[int, str] = (1, 'a')
+
+ParamAlias = TypeAliasType("ParamAlias", Callable[P, int], type_params=(P,))
+def f(x: str, y: float) -> int: return 1
+def g(x: int, y: float) -> int: return 1
+xp1: ParamAlias[str, float] = f
+xp2: ParamAlias[str, float] = g  # E: Incompatible types in assignment (expression has type "Callable[[int, float], int]", variable has type "Callable[[str, float], int]")
+xp3: ParamAlias[str, float] = lambda x, y: 1
+[builtins fixtures/dict.pyi]
+
+[case testTypeAliasTypeInvalidGeneric]
+from typing_extensions import TypeAliasType, TypeVarTuple, ParamSpec
+from typing import Dict, TypeVar, Tuple, Unpack
+
+K = TypeVar('K')
+V = TypeVar('V')
+Ts = TypeVarTuple("Ts")
+Ts1 = TypeVarTuple("Ts1")
+P = ParamSpec("P")
+
+Ta1 = TypeAliasType("Ta1", int, type_params=K)  # E: Tuple literal expected as the type_params argument to TypeAliasType
+
+Ta2 = TypeAliasType("Ta2", int, type_params=(None,))  # E: Free type variable expected in type_params argument to TypeAliasType
+
+Ta3 = TypeAliasType("Ta3", Dict[K, V], type_params=(V,))  # E: Can't use bound type variable "K" to define generic alias
```

Comment:
Not a very clear error. Can this instead say "Type variable K is not included in type_params" or something similar?

Reply:
I like this suggestion. Unless you have an alternative that would also cover the same case but for aliases written with the `type` keyword, I'll go with the your current suggestion.